### ISSUE TITLE
Add DictionaryKeyPolicy to CosmosSerializationOptions

### ISSUE THREAD

**Issue by davecluderay** [Azure/azure-cosmos-dotnet-v3#3810]
`System.Text.Json` allows consumers to specify a different naming policy for dictionary keys than for regular properties.
This is supported via a [DictionaryKeyPolicy](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/customize-properties?pivots=dotnet-7-0#camel-case-dictionary-keys) property on `JsonSerializerOptions`.
A similar capability should be provided for the out-of-the-box Cosmos serializer without having to implementing a custom serializer.

**Comment by ealsur** [comment]
@davecluderay The SDK uses Newtonsoft.Json as default serialization engine, is this available there?

**Comment by davecluderay** [comment]
@ealsur The same can be achieved with Newtonsoft.Json, though the configuration is obviously not identical.
It is supported via [NamingStrategy.ProcessDictionaryKeys](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_NamingStrategy_ProcessDictionaryKeys.htm).

**Comment by ealsur** [comment]
This property is set to true for the `CamelCasePropertyNamesContractResolver`. And the `CamelCasePropertyNamesContractResolver` is used if you set `CosmosPropertyNamingPolicy = CamelCase` in the `CosmosSerializationOptions`:

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/31b1ff330a4b0983f8e29a1156faaa0c702584d5/Microsoft.Azure.Cosmos/src/Serializer/CosmosJsonDotNetSerializer.cs#L48-L56

Doesn't that work as expected?

**Comment by ealsur** [comment]
As reference, this is the source code of `CamelCasePropertyNamesContractResolver`:

![image](https://user-images.githubusercontent.com/1633401/232148418-e3ceb073-0728-45e0-8e9a-2685fed4415b.png)

Newtonsoft.Json does not allow for the Dictionary policy to be different from regular properties, in that case, you'd need to use a Converter on the property itself.

**Comment by davecluderay** [comment]
This example specifically sets the `ProcessDictionaryKeys` to `false` on initialising the `CamelCaseNamingStrategy`: [NamingStrategySkipDictionaryKeys](https://github.com/JamesNK/Newtonsoft.Json/blob/0a2e291c0d9c0c7675d445703e51750363a549ef/Src/Newtonsoft.Json.Tests/Documentation/Samples/Serializer/NamingStrategySkipDictionaryKeys.cs#L73).

The `CosmosJsonDotNetSerializer` could initialise this property to `true` or `false` based on its own options I think?

**Comment by ealsur** [comment]
@davecluderay So your ask is to be able to disable `ProcessDictionaryKeys` even though you are setting `CosmosPropertyNamingPolicy = CamelCase`?

**Comment by davecluderay** [comment]
Yes, that is right. I want to be able to configure it so that dictionary keys pass through unchanged, regardless of the property naming strategy.

**Comment by ealsur** [comment]
I see two paths:

1. Add a bool to the `CosmosSerializationOptions` to apply the NamingPolicy to Dictionaries. The problem is mixing the boolean with a NamingPolicy that would not apply (like Default).
2. Add a new enum value to `CosmosPropertyNamingPolicy`, like `CamelCaseOnlyProperties`. This approach makes sure we set the flag in the `CamelCasePropertyNamesContractResolver` to `false`.

I'll mark this as feature request to be tracked.

**Comment by alex-ht-work** [comment]
This should be classified as a bug report, not a feature request.
Lowercasing of dictionary keys is a corruption of user data, which should never be done by any library or framework.
